### DISCUSSION TITLE
building an unstructured MF6 simulation from already existing structured model.

### DISCUSSION THREAD

**Discussion by sebmeng** [modflowpy/flopy#2322] [Q&A] [answered]
Hello community!
I would like to ask for some guidance. I have a structured mf6 model with several packages (chd, riv, npf, to name a few) that I loaded into flopy:
`sim = flopy.mf6.MFSimulation.load("MF", sim_ws=resultdir, strict=True, verbosity_level=2, verify_data=False)`

I used GRIDGEN to create an unstructured quadtree grid using the Gridgen module of flopy:

```
# Initialize the Gridgen object with the existing grid (changed executable to x64 to see if it works with two levels of refinement)
g = Gridgen(dis, model_ws=gridgen_dir,exe_name=r'C:\Users\926741\Executables\gridgen.1.0.02\bin\gridgen_x64.exe')

# Add the refinement area using the polygon shapefile
g.add_refinement_features([geom1], 'polygon', 1, (range(nlay)))  # 1 is the level of refinement


g.add_refinement_features([geom2], 'polygon', 2, (range(nlay)))  # 2 is the level of refinement


g.build(verbose=True)
```

and got a unstructured grid with two levels of refinement: 
![image](https://github.com/user-attachments/assets/e50804cd-9016-4296-ac42-d3e8cf03695a)

Now i want to "update" this structured simulations using this structured grid (is not possible for me to create the input for the untructured packages, but rather use the structured ones from this flopy sim object, and "convert" them into this new unstructured grid)

i thought that i could create a disv object with this piece of code:

```
disv_props = g.get_gridprops_disv()

disv = flopy.mf6.ModflowGwfdisv(
    sim, 
    nlay=disv_props['nlay'], 
    ncpl=disv_props['ncpl'], 
    nvert=disv_props['nvert'], 
    vertices=disv_props['vertices'], 
    cell2d=disv_props['cell2d'], 
    top=disv_props['top'], 
    botm=disv_props['botm'], 
    xorigin=grid.xoffset, 
    yorigin=grid.yoffset, 
    angrot=grid.angrot, 
    length_units=grid.lenuni,
)
```
but of course I cannot do it because "sim' is structured: 

`Invalid package type "disv".  Unable to register package.
`
Any  guidance on how to tackle this issue?

Appreciate any help,

Thanks!

Sebastian

**Comment by christianlangevin** [comment]
Hi @sebmeng, there is no way to simply convert your structured model into an unstructured disv model.  You will need to rebuild your simulation and gwf model packages (chd, riv, npf, etc) using the new grid.  We don't have an easy way to convert from one to the other.  There are decisions that you will have to make, such as, if there is a well in the structured model, how should that pumping be distributed in the unstructured model?  Fortunately, flopy has raster and shape intersection capabilities, so theoretically, you can write a script that can build a model for whatever grid you want, but that is something you need to design from the beginning.

**Reply by sebmeng** [reply]
Thanks for the quick reply @langevin-usgs 
I understand what you're saying, it is not that simple to do it.
But adding up to your suggestion, I could export the packages information of the structured grid as shapefiles or rasters and then build a new simulation with the new grid and use that raster/shape information from every package to build the corresponding package in this new unstructured siumulation, right?

**Reply by christianlangevin** [reply, marked as answer]
Sounds like you might be able to do what you want, but I think the best approach would be to use the native information in shapefiles and rasters and write a script to build either type of model.  But I recognize that may not be possible.

**Reply by sebmeng** [reply]
Thanks for the suggestion @langevin-usgs , I will go with that.
Best,
Sebastian
